src/scales/scale.time.js

```diff
@@ -364,42 +358,60 @@ function ticksFromTimestamps(scale, values, majorUnit) {
 	return (ilen === 0 || !majorUnit) ? ticks : setMajorTicks(scale, ticks, map, majorUnit);
 }
 
-
 function getDataTimestamps(scale) {
 	var timestamps = scale._cache.data || [];
 	var i, ilen, metas;
 
-	if (!timestamps.length) {
-		metas = scale._getMatchingVisibleMetas();
-		for (i = 0, ilen = metas.length; i < ilen; ++i) {
-			timestamps = timestamps.concat(metas[i].controller._getAllParsedValues(scale));
-		}
-		timestamps = scale._cache.data = arrayUnique(timestamps).sort(sorter);
+	if (timestamps.length) {
+		return timestamps;
 	}
-	return timestamps;
+
+	metas = scale._getMatchingVisibleMetas();
+	for (i = 0, ilen = metas.length; i < ilen; ++i) {
+		timestamps = timestamps.concat(metas[i].controller._getAllParsedValues(scale));
+	}
+
+	// We can not assume data is in order or unique - not even for single dataset
```

Comment:
We have always assumed that data is in order and unique for a single dataset. I'm not sure I want to change that. Both because I don't know what other code has those assumptions built in and also because it'd be worse from a performance point of view

Reply:
Not always, this assumption was made in quite recent optimizations. Originally `arrayUnique` was called regardless of number of datasets.

I can think of use cases where timestamps would not be in order (and should be drawn that way). But those are not the "usual" cases, so maybe it should be behind a switch.

I think this would be a lot more maintainable, if "series" and "linear" time scales were separated. "linear" could be based mostly on `LinearBase`.

Reply:
I think I agree that it'd be more maintainable to have "series" and "linear" as separate scales. There's a lot of complication trying to cram both concepts into a single scale. I'm not sure there's a lot you could share between the linear time scale and linear scale. There is a lot you could share between the linear and series time scales though. Not sure whether it'd be better to have one extend the other or make a base time scale

Reply:
I'll explore that in a later PR